File: src/main/java/com/jcloisterzone/feature/RangeFeature.java
package com.jcloisterzone.feature;

import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.game.state.PlacedTile;
import io.vavr.collection.Stream;

public interface RangeFeature extends Feature {

    Stream<PlacedTile> getRangeTiles(GameState state);
    
    default Stream<PlacedTile> getRangeTilesWithFeature(GameState state) {
        Stream<PlacedTile> featureTiles = getPlaces()
            .toStream()
            .map(fp -> fp.getPosition())
            .distinct()
            .map(pos -> state.getPlacedTiles().get(pos))
            .flatMap(opt -> opt.toStream());

        return getRangeTiles(state).appendAll(featureTiles);
    }

}

File: src/main/ts/com/jcloisterzone/feature/RangeFeature.ts
import type { Stream } from "../../../io/vavr/SeqTypes.js";
import type { GameState } from "../game/state/GameState.js";
import type { PlacedTile } from "../game/state/PlacedTile.js";
import type { Feature } from "./Feature.js";

/** A feature with a range of tiles (cloister/garden/dragon vicinity). */
export interface RangeFeature extends Feature {
  getRangeTiles(state: GameState): Stream<PlacedTile>;
  /** default: range tiles plus the feature's own tiles. */
  getRangeTilesWithFeature(state: GameState): Stream<PlacedTile>;
}

/** Runtime mirror of Java's `instanceof RangeFeature` (duck-typed on
 *  getRangeTilesWithFeature). THE single definition. */
export function isInstanceOfRangeFeature(f: unknown): f is RangeFeature {
  return typeof (f as { getRangeTilesWithFeature?: unknown } | null)?.getRangeTilesWithFeature === "function";
}
